rust/osm2lanes/src/metric.rs

```diff
@@ -0,0 +1,330 @@
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Metre(f64);
+
+impl Metre {
+    pub const fn new(val: f64) -> Self {
+        Self(val)
+    }
+    pub const fn val(&self) -> f64 {
+        self.0
+    }
+}
+
+impl std::ops::Add for Metre {
+    type Output = Self;
+    fn add(self, other: Self) -> Self {
+        Self(self.0 + other.0)
+    }
+}
+impl std::ops::AddAssign for Metre {
+    fn add_assign(&mut self, other: Self) {
+        *self = Self(self.0 + other.0);
+    }
+}
+impl std::ops::Mul<Metre> for f64 {
+    // The division of rational numbers is a closed operation.
+    type Output = Metre;
+    fn mul(self, other: Metre) -> Self::Output {
+        Metre::new(self * other.val())
+    }
+}
+impl std::iter::Sum for Metre {
+    fn sum<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = Metre>,
+    {
+        Self(iter.map(|m| m.0).sum())
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Speed {
+    Kph(f64),
```

Comment:
An enum is harder for an external user to deal with. Why don't we convert everything to kph, like distance? Also I think it vastly simplifies all of the serialization code!

Reply:
A renderer may/will want to show the speed limit per lane? And this visualisation will need to know the locale of how this is displayed, and this is know by osm2lanes (and hopefully not so much by the rendered)?

I would have preferred that as a simpler approach as well

Reply:
I just realised I should have documented my whole attempt at making this an unisgned integer.

but then I discovered the amazing world of US fractional speed limits :/

Reply:
Good point, locale does matter for rendering. I wonder if it'd be simpler to give people a way to format a `Speed` struct with the locale for that use case. It makes the cross-language API bindings harder, though, since we're exposing methods on more structs.

I'd still vote for canonical units, but this approach is fine by me too

Reply:
I guess most renderers will need to know how to render the signs differently anyway, so they could convert back to mph?

idk

Reply:
maybe a struct like 

```rust
Speed {
    unit: Unit, # e.g. "mph"
    kph: f64,
}

Reply:
I think I will leave like this now indeed.

It might be hard (if they use the enum, not the flattened serialized API), but at least it has few surprised as it mirrors OSM's maxspeed tag.

Reply:
Interesting idea. Why's that need to be repeated for every Speed value though? Why not just do `locale.speed_unit() -> Unit`?

Reply:
oh yeah, true.

I guess the the speed unit (along with driving side, and other locale details), would sit at the same level as highway and lifecycle does, outside the lanes.

Reply:
then you can just do `speed: f64` and have the data consumer use the global locale information...

I would just need to check that there isn't some edge case where cars have a speed limit in mph, but busses, trams, or bicycles have one in kph for example... that would suck.